In [71]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [72]:
df = pd.read_csv('wk12.csv')

In [73]:
df

Wk      Time            Winner  Pts             Loser  Pts.1  \
0     1   7:00 PM  Florida Atlantic   43         Charlotte     13   
1     1   5:00 PM     Florida State   47          Duquesne      7   
2     1   4:00 PM          Illinois   38           Wyoming      6   
3     1  10:00 PM            Nevada   23  New Mexico State     12   
4     1   3:30 PM  Nevada-Las Vegas   52       Idaho State     21   
..   ..       ...               ...  ...               ...    ...   
709  12   7:00 PM   (24) Washington   37        (6) Oregon     34   
710  12   3:30 PM  Washington State   28     Arizona State     18   
711  12  12:00 PM     West Virginia   23          Oklahoma     20   
712  12   2:00 PM  Western Kentucky   45              Rice     10   
713  12   7:00 PM           Wyoming   14    Colorado State     13   

              winners            losers  home_rank  away_rank  home_team_code  \
0    Florida Atlantic         Charlotte         99         99              32   
1       Florida State          Duquesne         99         99              34   
2            Illinois           Wyoming         99         99              42   
3              Nevada  New Mexico State         99         99              76   
4    Nevada-Las Vegas       Idaho State         99         99              74   
..                ...               ...        ...        ...             ...   
709        Washington            Oregon          6         24              88   
710  Washington State     Arizona State         99         99             126   
711     West Virginia          Oklahoma         99         99             127   
712  Western Kentucky              Rice         99         99             128   
713           Wyoming    Colorado State         99         99              26   

     away_team_code  home_results  win_rank  lose_rank       dates  day_code  
0                37             0        99         99  2022-08-27         5  
1                49             0        99         99  2022-08-27         5  
2               221             0        99         99  2022-08-27         5  
3               121             1        99         99  2022-08-27         5  
4                76             0        99         99  2022-08-27         5  
..              ...           ...       ...        ...         ...       ...  
709             210             1        24          6  2022-11-12         5  
710              11             0        99         99  2022-11-12         5  
711             143             0        99         99  2022-11-12         5  
712             153             0        99         99  2022-11-12         5  
713             221             1        99         99  2022-11-12         5  

[714 rows x 17 columns]

In [74]:
df = df.drop(columns=['Time', 'Winner', 'Loser', 'dates'])

In [75]:
df

Wk  Pts  Pts.1           winners            losers  home_rank  away_rank  \
0     1   43     13  Florida Atlantic         Charlotte         99         99   
1     1   47      7     Florida State          Duquesne         99         99   
2     1   38      6          Illinois           Wyoming         99         99   
3     1   23     12            Nevada  New Mexico State         99         99   
4     1   52     21  Nevada-Las Vegas       Idaho State         99         99   
..   ..  ...    ...               ...               ...        ...        ...   
709  12   37     34        Washington            Oregon          6         24   
710  12   28     18  Washington State     Arizona State         99         99   
711  12   23     20     West Virginia          Oklahoma         99         99   
712  12   45     10  Western Kentucky              Rice         99         99   
713  12   14     13           Wyoming    Colorado State         99         99   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
0                32              37             0        99         99   
1                34              49             0        99         99   
2                42             221             0        99         99   
3                76             121             1        99         99   
4                74              76             0        99         99   
..              ...             ...           ...       ...        ...   
709              88             210             1        24          6   
710             126              11             0        99         99   
711             127             143             0        99         99   
712             128             153             0        99         99   
713              26             221             1        99         99   

     day_code  
0           5  
1           5  
2           5  
3           5  
4           5  
..        ...  
709         5  
710         5  
711         5  
712         5  
713         5  

[714 rows x 13 columns]

In [76]:
df[df['winners'] == 'Oregon']

Wk  Pts  Pts.1 winners              losers  home_rank  away_rank  \
144   3   70     14  Oregon  Eastern Washington         99         99   
222   4   41     20  Oregon       Brigham Young         25         12   
291   5   44     41  Oregon    Washington State         99         15   
359   6   45     27  Oregon            Stanford         13         99   
414   7   49     22  Oregon             Arizona         99         12   
520   9   45     30  Oregon                UCLA         10          9   
579  10   42     24  Oregon          California         99          8   
628  11   49     10  Oregon            Colorado         99          8   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
144              88              54             0        99         99   
222              88              24             0        25         12   
291             126             146             1        15         99   
359              88             177             0        13         99   
414               5             146             1        12         99   
520              88             200             0        10          9   
579              18             146             1         8         99   
628              25             146             1         8         99   

     day_code  
144         5  
222         5  
291         5  
359         5  
414         5  
520         5  
579         5  
628         5

In [77]:
df[df['winners'] == 'Utah']

Wk  Pts  Pts.1 winners               losers  home_rank  away_rank  \
168   3   73      7    Utah        Southern Utah         13         99   
238   4   35      7    Utah      San Diego State         14         99   
311   5   34     13    Utah        Arizona State         99         13   
372   6   42     16    Utah         Oregon State         12         99   
483   8   43     42    Utah  Southern California         20          7   
543  10   21     17    Utah     Washington State         99         14   
644  11   45     20    Utah              Arizona         12         99   
706  12   42      7    Utah             Stanford         13         99   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
168             119             175             0        13         99   
238             119             160             0        14         99   
311               6             201             1        13         99   
372             119             147             0        12         99   
483             119             171             0        20          7   
543             126             201             1        14         99   
644             119              10             0        12         99   
706             119             177             0        13         99   

     day_code  
168         5  
238         5  
311         5  
372         5  
483         5  
543         3  
644         5  
706         5

In [78]:
# imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as se
import warnings
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
warnings.filterwarnings('ignore')
from sklearn.preprocessing import MinMaxScaler
from sklearn.feature_selection import SelectKBest,f_regression,f_classif
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import PowerTransformer
from sklearn.ensemble import GradientBoostingRegressor

In [79]:
df.head()

Wk  Pts  Pts.1           winners            losers  home_rank  away_rank  \
0   1   43     13  Florida Atlantic         Charlotte         99         99   
1   1   47      7     Florida State          Duquesne         99         99   
2   1   38      6          Illinois           Wyoming         99         99   
3   1   23     12            Nevada  New Mexico State         99         99   
4   1   52     21  Nevada-Las Vegas       Idaho State         99         99   

   home_team_code  away_team_code  home_results  win_rank  lose_rank  day_code  
0              32              37             0        99         99         5  
1              34              49             0        99         99         5  
2              42             221             0        99         99         5  
3              76             121             1        99         99         5  
4              74              76             0        99         99         5

In [80]:
fghjk

NameError: name 'fghjk' is not defined

In [81]:
X = df.drop(columns=['winners', 'losers', 'Pts.1'])

In [82]:
y = df['Pts.1']

In [83]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=123)


In [84]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression

In [85]:
regr = RandomForestRegressor(max_depth=50000, random_state=0)
regr.fit(X_train, y_train)

RandomForestRegressor(max_depth=50000, random_state=0)

In [86]:
test_pred = regr.predict(X_test)

print('RMSE: '+str(np.sqrt(mean_squared_error(y_test, test_pred))))
print('R Squared: '+str(r2_score(y_test, test_pred)))
print('MSE TEST: '+str(mean_squared_error(y_test, test_pred)))
print('MAE TEST: '+str(mean_absolute_error(y_test, test_pred)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, test_pred)))

RMSE: 10.476451749458938
R Squared: -0.06409468924887585
MSE TEST: 109.75604125874126
MAE TEST: 8.266923076923074
MAPE: 3347150827950604.0


In [87]:
df[df['winners'] == 'Utah']

Wk  Pts  Pts.1 winners               losers  home_rank  away_rank  \
168   3   73      7    Utah        Southern Utah         13         99   
238   4   35      7    Utah      San Diego State         14         99   
311   5   34     13    Utah        Arizona State         99         13   
372   6   42     16    Utah         Oregon State         12         99   
483   8   43     42    Utah  Southern California         20          7   
543  10   21     17    Utah     Washington State         99         14   
644  11   45     20    Utah              Arizona         12         99   
706  12   42      7    Utah             Stanford         13         99   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
168             119             175             0        13         99   
238             119             160             0        14         99   
311               6             201             1        13         99   
372             119             147             0        12         99   
483             119             171             0        20          7   
543             126             201             1        14         99   
644             119              10             0        12         99   
706             119             177             0        13         99   

     day_code  
168         5  
238         5  
311         5  
372         5  
483         5  
543         3  
644         5  
706         5

utah 10 at oregeon 12
utah away code = 201
oregon home code = 146

In [88]:
df[df['winners'] == 'Oregon']

Wk  Pts  Pts.1 winners              losers  home_rank  away_rank  \
144   3   70     14  Oregon  Eastern Washington         99         99   
222   4   41     20  Oregon       Brigham Young         25         12   
291   5   44     41  Oregon    Washington State         99         15   
359   6   45     27  Oregon            Stanford         13         99   
414   7   49     22  Oregon             Arizona         99         12   
520   9   45     30  Oregon                UCLA         10          9   
579  10   42     24  Oregon          California         99          8   
628  11   49     10  Oregon            Colorado         99          8   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
144              88              54             0        99         99   
222              88              24             0        25         12   
291             126             146             1        15         99   
359              88             177             0        13         99   
414               5             146             1        12         99   
520              88             200             0        10          9   
579              18             146             1         8         99   
628              25             146             1         8         99   

     day_code  
144         5  
222         5  
291         5  
359         5  
414         5  
520         5  
579         5  
628         5

In [89]:
X.head()

Wk  Pts  home_rank  away_rank  home_team_code  away_team_code  \
0   1   43         99         99              32              37   
1   1   47         99         99              34              49   
2   1   38         99         99              42             221   
3   1   23         99         99              76             121   
4   1   52         99         99              74              76   

   home_results  win_rank  lose_rank  day_code  
0             0        99         99         5  
1             0        99         99         5  
2             0        99         99         5  
3             1        99         99         5  
4             0        99         99         5

In [90]:
on = [[13, 41, 12, 10, 146, 201, 0, 12, 10, 5]]
uo = pd.DataFrame(on, columns=['Wk', 'Pts', 'home_rank', 'away_rank',
                                'home_team_code', 'away_team_code', 'home_results', 
                               'win_rank', 'lose_rank', 'day_code'])
uo

Wk  Pts  home_rank  away_rank  home_team_code  away_team_code  \
0  13   41         12         10             146             201   

   home_results  win_rank  lose_rank  day_code  
0             0        12         10         5

In [91]:
test_pred = regr.predict(uo)
test_pred

array([31.27])

# oregon wins 0.45

## Oregon 40.6 - Utah 31.27

## Oddshark prediction 
oregon 41.2 -  Utah 36.8

spread is oregon -3
over/under is 61.5

###  R-Squared reading can be much higher, such as 0.9 or above. In finance, an R-Squared above 0.7 would generally be seen as showing a high level of correlation, whereas a measure below 0.4 would show a low correlation.

### The lower the MAE score the better. This is because MAE is a measure of the average error between the predictions and intended targets, thus we want to minimise this value.

### The lower the MSE value there more accurate the model is. 